src/cascade/stats/estimation.py

```diff
@@ -47,3 +88,61 @@ def bounds_to_stdev(lower, upper):
         pd.Series: The standard deviation for that CI.
     """
     return (upper - lower) / (2 * 1.96)
+
+
+def stdev_from_bundle_data(bundle_df):
+    """
+    Takes a bundle data frame and figures out the standard deviation
+    from what is included in the bundle.
+
+    There are other functions that will still use the bounds_to_stdev
+    function rather than this because they're not dealing with
+    bundle data. This function should only be used for bundle data.
+
+    We prefer standard deviation (has_se), then uncertainty intervals (has_ui),
+    then effective sample size (has_es), then sample size (has_ss).
+
+    Args:
+        bundle_df:
+
+    Returns:
+
+    """
+    df = bundle_df.copy()
+    standard_error = df.standard_error.values.copy()
+
+    has_se = ~df.standard_error.values.isnull() & df.standard_error.values > 0
+    MATHLOG.info(f"{sum(has_se)} rows have standard error.")
+    has_ui = ~df.lower.values.isnull() & ~df.upper.values.isnull()
+    MATHLOG.info(f"{sum(has_ui)} rows have uncertainty.")
+    has_es = ~df.effective_sample_size.values.isnull() & df.effective_sample_size.values > 0
+    MATHLOG.info(f"{sum(has_es)} rows have effective sample size.")
+    has_ss = ~df.sample_size.values.isnull() & df.sample_size.values > 0
+    MATHLOG.info(f"{sum(has_ss)} rows have sample size.")
+
+    if sum(has_se | has_ui | has_es | has_ss) < len(df):
+        raise ValueError("Some rows have no valid uncertainty.")
+
+    replace_es_with_ss = ~has_es & has_ss
+    MATHLOG.info(f"{sum(replace_es_with_ss)} rows will have their effective sample size filled by sample size.")
+    replace_se_with_ui = ~has_se & has_ui
+    MATHLOG.info(f"{sum(replace_se_with_ui)} rows will have their standard error filled by uncertainty intervals.")
+    replace_se_with_es = ~has_se & ~has_ui
+    MATHLOG.info(f"{sum(replace_se_with_es)} rows will have their standard error filled by effective sample size.")
+
+    # Replace effective sample size with sample size
+    df[replace_es_with_ss, 'effective_sample_size'] = df[replace_es_with_ss, 'sample_size']
+
+    # Calculate standard deviation different ways (also
+    stdev_from_bounds = bounds_to_stdev(lower=df.lower.values, upper=df.upper.values)
+    stdev_from_es = es_to_stdev(mean_value=df.mean.values, es=df.effective_sample_size.values)
+
+    # Use boolean arrays representing the pecking order to replacing standard error
+    standard_error[replace_se_with_ui] = stdev_from_bounds[replace_se_with_ui]
+    standard_error[replace_se_with_es] = stdev_from_es[replace_se_with_es]
+
+    # Do a final check on standard error
```

Comment:
I am wondering if we should allow std's == 0 to pass through, and let dismod fix them with minCV by integrand. If we decide to do that, we should catch the case where both mean == std == 0. That will stop dismod.

Reply:
I think this is something to flag with the team. Left a comment on the JIRA ticket DISMODAT-544 for us to figure out.